README.md
<!--
 * @Author: Drazy
 * @Date: 2021-09-02 14:50:29
 * @LastEditors: Drazy
 * @LastEditTime: 2022-01-19 17:31:36
 * @Description: Readme
-->
# 设计模式 案例集合

#### 这是一个将 React 与 TypeScript 相结合去学习设计模式的过程

各个版本的详细细节见 [更新日志](./CHANGELOG.md)

---

**主要包括的学习部分：**

 1. 简单工厂模式
 2. 策略模式
 3. 原型模式
 4. 装饰模式
 5. 单例模式
 6. 观察者模式
 7. 责任链模式
 8. 桥接模式
 9. 命令模式
 10. 备忘录模式
 11. 工厂模式
 12. 抽象工厂模式
 13. 适配器模式
 14. 代理模式
 15. 模板方法模式
 16. 外观/门面模式
 17. 建造者模式
 18. 组合模式
 19. 状态模式
 20. 迭代器模式
 21. 中介者模式
 22. 享元模式
 23. 解释器模式
 24. 访问者模式


CHANGELOG.md
# 更新日志

## 2.0.0(2022.01.19)

### **解释器**模式

实现了 读懂老板的心案例、音乐解释器案例

1. 用处：

   - 如果一种特定类型的问题发生的频率足够高，那么可能就值得将该问题的各个实例表述为一个简单语言中的句子。通过构建一个解释器，利用解释器来解释这些句子从而解决问题

   - 当有一个语言需要解释执行，并且可以将该语言的句子表示为一个抽象语法树，可以使用解释器模式

2. 特点：

   - 解释器模式为文法中的每一条规则都至少定义了一个类，因此包含了许多规则的文法难以管理和维护。建议当文法非常复杂时，使用其他的技术（如语法分析程序或者编译器生成器）来处理

3. 好处：

   - 容易改变和扩展待解释语言的文法，比较容易实现文法

### **访问者**模式

实现了 男人女人语录案例

1. 用处：

   - 如果一个系统具有比较稳定的数据结构，又有易于变化的算法，可以考虑使用访问者模式。因为访问者模式使得新增算法操作的行为变得较为容易

2. 特点：

   - 将数据结构和作用于结构之上的操作间的耦合解脱开，使得操作集合可以相对自由地进行演化

   - 可以使得在不改变各元素的类的前提下定义作用于这些元素的新操作

3. 好处：

   - 增加新的操作很容易，因为增加一个新的操作就意味着增加一个新的访问者。访问者模式实际上是将有关的行为集中到一个访问者对象之中

4. 坏处：

   - 使得增加新的数据结构变得较为困难（每增加一个新的状态，就需要在所有的具体访问者中都新增一个方法，这就不符合开闭原则了）

---

## 1.9.9(2021.12.27)

### **享元**模式

实现了 网站共享代码案例

1. 用处：

   - 如果一个应用程序使用了搭理那个的对象，而大量的这些对象造成了很大的存储开销时就应该考虑使用享元模式

   - 以及对象的大多数状态可以外部状态，如果删除对象的外部状态，那么可以使用相对较少的共享对象取代很多组对象，此时也可以考虑使用享元模式

2. 特点：

   - 享元模式即 共享元素的模式 --> 共享对象的模式

   - 享元模式可以使实例总数大量减少，如果共享的对象越多，那么存储节约的也就越多，节约量可随着共享状态的增多而增大

   - 内部状态与外部状态
     - 内部状态：在享元对象内部并且不会随环境改变而改变的共享部分
     - 外部状态：随环境改变而改变，不可以共享的状态

3. 好处：

   - 可以避免大量非常相似类的开销

---

## 1.9.8(2021.11.09)

### **中介者**模式

实现了 联合国对话案例

1. 用处：
   一般在一组对象以定义良好但是复杂的方式进行通信时；以及想要定制一个分布在多个类中的行为，但又不想生成太多子类时； 可以考虑使用中介者模式

2. 特点：

   - 中介者模式很容易在系统中应用，也很容易在系统中误用。当系统中出现了‘多对多’的复杂对象群时，不要急于使用中介者模式，而是首先要反思你的系统在设计上是否合理

   - 中介者模式的缺点也是显而易见，由于中介者对象的功能过于集中化，也就是将交互的复杂性变成了中介者的复杂性，这会很容易让中介者对象变得比任何一个具体同事对象都更加复杂

3. 好处：
   通过中介者对象可以将系统中的网状结构变成以中介者为中心的星型结构，每个具体对象不再通过直接的联系与另一个对象发生相互作用，而是通过中介者对象与另一个对象发生相互作用

---

## 1.9.7(2021.11.02)

### **迭代器**模式

实现了 买票案例

1. 用处：
   当你需要访问一个聚集对象，并且对聚集有多种方式遍历的时候，就应该考虑使用迭代器模式

2. 特点：

   - 迭代器模式为遍历不同的聚集结构提供了如开始、下一个、是否结束、当前某一项等统一的接口

   - 迭代器模式在访问数组、集合、列表等数据时，尤其是对于数据库的数据操作时，是非常普遍的应用，但由于它用处太普遍了，所以各种高级语言也都对他进行了封装，因此反倒是给人感觉其本身不是很常用了

3. 好处：
   通过抽出一个迭代器类来分离集合对象的遍历行为，这样既可以不暴露集合的内部结构，又可以让外部代码透明地访问集合内部的数据

---

## 1.9.6(2021.10.12)

### **状态**模式

实现了 打工人不同时段工作状态案例

1. 用处：
   当一个对象的行为取决于它的状态，并且它必须在运行时时刻根据状态改变它的行为时，就可以考虑使用状态模式了

2. 特点：

   - 当一个对象的内在状态改变时允许改变其行为，这个对象看起来像是改变了其类

   - 主要解决的是当控制一个对象状态转换的条件表达式过于复杂时的情况（例如大量的 `if else` 语句）。将状态的判断逻辑转移到表示不同状态的一系列类之中，可以把复杂的判断逻辑简化

   - 通过把各种状态转移逻辑分布到 State 的子类之间，来减少相互间的依赖（开闭原则）

3. 好处：
   将与特定状态相关的行为局部化，并且将不同状态的行为分割开来

---

## 1.9.5(2021.09.06)

### **组合**模式

实现了 公司树形结构管理案例

1. 用处：
   需求中是体现部分与整体层次的结构时，以及你希望用户可以忽略组合对象与单个对象的不同，统一地使用组合结构中的所有对象时，就应该使用组合模式啦

2. 特点：

   - 组合模式定义了包含基本对象和组合对象的类层次结构

   - 基本对象可以被组合成为更复杂的组合对象，而这个对象又可以被组合，这样子不断递归下去，最后在应用时，任何用到基本对象的地方就都可以使用组合对象了

   - 组合模式在应用时可以一致性地使用组合结构和单个对象

---

## 1.9.4(2021.09.03)

### **建造者**模式

实现了 创建胖、瘦小人案例

1. 用处：

   主要用于创建一些复杂的对象，这些对象的内部构建间的建造顺序通常是稳定的，但对象内部的构建通常面临着复杂的变化

2. 特点：

   - 使得建造代码与表示代码分离，使同样的建造过程可以创建不同的表示

   - 如果需要改变一个产品的内部表示，只需要再定义一个具体的建造者就可以了

   - 建造者模式是在当在创建复杂对象的算法应该独立于对象的组成部分以及它们的装配方式时适用的模式

---

## 1.9.3(2021.06.10)

增添了 观察者设计模式 中 气象站发布新消息，诸多公告牌进行更新的案例

---

## 1.9.2(2021.05.06)

---

### **模板方法**模式

实现了 学生抄写试题案例

1. 用处：

   当不可变和可变的行为在方法的子类实现中混合在一起的时候，不变的行为就会在子类中重复出现。通过模板方法模式把这些行为搬移到单一的地方，这样就可以帮助子类摆脱重复不变的行为的纠缠。

2. 特点：

   通过把不变的行为搬移到抽象类中，去除子类中的重复代码来体现其优势

3. 感悟：

   感到非常实用，天天在用，平时对于公用逻辑或者业务在抽取抽象类的过程就是这个设计模式的实现

### **外观/门面**模式

实现了 股民炒股案例

1. 用处：

   - 在层与层之间可以建立外观模式。
   - 在开发时，子系统往往会因为不断重构演化而变得越来越复杂，增加一个外观类可以提供一个简单的接口，减少它们之间的依赖。
   - 在维护一个遗留的大型系统时，也可以为新系统开发一个外观类，来提供设计粗糙或高度复杂的遗留代码的比较清晰简单的接口，让新系统与外观对象交互，外观对象与遗留代码交互所有复杂的工作。

2. 特点：

   - 为子系统中的一组接口提供一个一致的界面，此模式定义了一个高层接口，这个接口使得这一子系统更加容易使用。

3. 感悟：

   - 外观和适配器模式都可以包装许多类。
   - 但外观的意图是简化接口，而适配器的意图是将接口转换成不同的接口

---

## 1.9.2(2021.04.30)

---

### **代理**模式

实现了 追求女生的案例

1. 用处：

   远程代理、虚拟代理、安全代理、智能指引

2. 特点：

   将其他对象提供一种代理以控制对这个对象的访问

---

## 1.9.1(2021.04.28)

---

### **适配器**模式

实现了 教练叫暂停后计划对战策略案例

1. 用处：

   当原始的某个方法或者类不方便改动，但又需要新的逻辑的时候，可以考虑使用适配器模式

2. 特点：

   将需要被适配的东西通过适配器转换为被需要的东西

3. 感悟：

   感到非常实用，平时的 Model 层其实就是对 api 接口与 view 页面所需数据的适配

### **代理**模式

实现了 追求女生的案例

---

## 1.9.0(2019.11.21)

---

1.  **工厂**模式
    >     实现了 生产不同厂家鼠标案例
    >     (1) 用处：
    >         当产品对象需要进行不同的加工时可以考虑使用 工厂方法
    >     (2) 特点：
    >         把工厂抽象出来，让子工厂来决定怎么生产产品， 每个产品都由自己的工厂生产
    >         符合开闭原则
    >     (3) 注意：
    >         这不是所谓的简单工厂模式的升级版，两者有不同的应用场景
2.  **抽象工厂**模式
    >     实现了 生产不同厂家不同硬件案例
    >     (1) 用处：
    >         当需要生产的是一个产品簇，并且产品之间或多或少有关联时可以考虑使用 抽象工厂方法
    >     (2) 特点：
    >         同样隐藏了具体产品的生产，不过生产的是多种类产品
    >     (3) 注意：
    >         和工厂模式的区别：工厂模式是一个产品，而抽象工厂模式是产品簇，是线和面的区别

## 1.8.0(2019.11.20)

---

1.  **备忘录**模式
    >     实现了 游戏存，读档案例
    >     (1) 用处：
    >         当对象需要保存/加载某一时刻的状态时（如游戏的save/load）可以考虑使用 备忘录模式
    >     (2) 特点：
    >         通过保存对象之前的状态来使对象可以恢复到之前的样子
    >         又称为 快照模式
    >         备忘录模式与命令模式常常可以一同使用
    >     (3) 实现方式：
    >         发起人（Originator) ： 依赖于 Memento；负责创建一个 Memento，用来记录当前时刻它的内部状态，并可以使用 Memento 恢复内部状态；可根据需要决定 Memento 存储 Originator 的哪些内部状态
    >         备忘录（Memento) ： 负责存储 Originator 的内部状态；可以防止除 Originator 以外的其他类访问 Memento；
    >         管理者（Caretaker) ： 聚合了 Memento；负责保存 Memento

## 1.7.0(2019.11.18)

---

1.  **命令**模式

    >     实现了 餐厅点餐案例
    >     (1) 用处：
    >         当需要发送者和接收者解耦时（常用于事件响应等）可以考虑使用 命令模式
    >     (2) 特点：
    >         把请求封装成命令对象，命令对象里包含有接收者，这样 client 只需要发送命令，接收者就可以做出相关响应或相反的响应
    >         将行为请求和行为实现解耦
    >         新的命令可以很容易添加到系统中去
    >         可能会导致有过多的具体命令类
    >     (3) 实现方式：
    >         请求者类（Invoker) ： 聚合了 Command；负责调用 Command 对象发起执行
    >         命令抽象类（Command) ： 用来声明执行操作的接口
    >         具体命令类（ConcreteCommand) ： 关联了 Receiver；将一个 Receiver 对象绑定于一个动作，调用 Receiver 中相应的操作，来实现 Command 中定义的操作
    >         接受者类（Receiver) ： 负责请求具体内容的执行

2.  实现了 游戏存，读档案例
    >     希望能够完善读存，读档的方法

## 1.6.1(2019.11.12)

---

1.  实现了 餐厅点餐案例
    >     希望能够改善顾客，服务员，订单，厨师之间的高耦合情况

## 1.6.0(2019.11.11)

---

1.  **责任链**模式

    >     实现了 员工请假案例
    >     (1) 用处：
    >         当请求需要被链式处理时，且每次事件的处理者不一定时（比如事件的冒泡等），可以考虑使用 责任链模式
    >     (2) 特点：
    >         无须知道是哪个具体的处理者去处理的请求，降低了耦合度
    >         请求处理对象时，仅需要维持一个指向其后继者的引用
    >         增加或删除一个责任链时，只需要重新建立链条即可。责任链的灵活度很高
    >     (3) 实现方式：
    >         抽象 处理者（Handler) ： 定义一个处理请示的接口，需要包含可以设置后继处理者的方法和能够处理请求的抽象方法

src/SiderMenu.ts
import { RouterEnum } from './router/RouterEnum'

/*
 * @Author: Drazy
 * @Date: 2021-11-10 14:31:44
 * @LastEditors: Drazy
 * @LastEditTime: 2022-01-19 15:22:29
 */
export interface IMenuItem {
  linkTo: RouterEnum
  Icon: string
  text: string
  exampleName?: string
}

export interface ISubMenu {
  title: IMenuItem
  menuItemList: IMenuItem[]
}

export interface ISider {
  isSubMenu: boolean
  linkTo?: RouterEnum
  Icon?: string
  text?: string
  subMenu?: ISubMenu
}

export const siderMenuList: ISider[] = [
  {
    isSubMenu: false,
    linkTo: RouterEnum.home,
    Icon: 'home',
    text: '介绍',
  },
  {
    isSubMenu: true,
    subMenu: {
      title: {
        linkTo: RouterEnum.factory,
        Icon: 'rocket',
        text: '创建型模式',
      },
      menuItemList: [
        {
          linkTo: RouterEnum.factory,
          Icon: 'shop',
          text: '工厂模式',
          exampleName: '生产不同厂家鼠标案例',
        },
        {
          linkTo: RouterEnum.abstractFactory,
          Icon: 'shop',
          text: '抽象工厂模式',
          exampleName: '生产不同厂家不同硬件案例',
        },
        {
          linkTo: RouterEnum.simpleFactory,
          Icon: 'shop',
          text: '简单工厂模式',
          exampleName: '计算器案例',
        },
        {
          linkTo: RouterEnum.singleton,
          Icon: 'minus',
          text: '单例模式',
          exampleName: '太阳月亮案例',
        },
        {
          linkTo: RouterEnum.builder,
          Icon: 'scissor',
          text: '建造者模式',
          exampleName: '创建胖、瘦小人案例',
        },
        {
          linkTo: RouterEnum.prototype,
          Icon: 'import',
          text: '原型模式',
          exampleName: '克隆人案例',
        },
      ],
    },
  },
  {
    isSubMenu: true,
    subMenu: {
      title: {
        linkTo: RouterEnum.adapater,
        Icon: 'database',
        text: '结构型模式',
      },
      menuItemList: [
        {
          linkTo: RouterEnum.adapater,
          Icon: 'api',
          text: '适配器模式',
          exampleName: '教练叫暂停后计划对战策略案例',
        },
        {
          linkTo: RouterEnum.bridge,
          Icon: 'column-width',
          text: '桥接模式',
          exampleName: '手机安装软件案例',
        },
        {
          linkTo: RouterEnum.combination,
          Icon: 'apartment',
          text: '组合模式',
          exampleName: '公司树形结构管理案例',
        },
        {
          linkTo: RouterEnum.decorator,
          Icon: 'border-outer',
          text: '装饰模式',
          exampleName: '人穿衣服案例',
        },
        {
          linkTo: RouterEnum.facade,
          Icon: 'skin',
          text: '外观/门面模式',
          exampleName: '股民炒股案例',
        },
        {
          linkTo: RouterEnum.agency,
          Icon: 'block',
          text: '代理模式',
          exampleName: '追求女生案例',
        },
        {
          linkTo: RouterEnum.flyweight,
          Icon: 'fork',
          text: '享元模式',
          exampleName: '网站共享代码案例',
        },
      ],
    },
  },
  {
    isSubMenu: true,
    subMenu: {
      title: {
        linkTo: RouterEnum.chainOfResponsibility,
        Icon: 'setting',
        text: '行为型模式',
      },
      menuItemList: [
        {
          linkTo: RouterEnum.chainOfResponsibility,
          Icon: 'cluster',
          text: '责任链模式',
          exampleName: '员工请假案例',
        },
        {
          linkTo: RouterEnum.command,
          Icon: 'code',
          text: '命令模式',
          exampleName: '餐厅点餐案例',
        },
        {
          linkTo: RouterEnum.iterator,
          Icon: 'retweet',
          text: '迭代器模式',
          exampleName: '买票案例',
        },
        {
          linkTo: RouterEnum.mediator,
          Icon: 'smile-o',
          text: '中介者模式',
          exampleName: '联合国对话案例',
        },
        {
          linkTo: RouterEnum.memento,
          Icon: 'container',
          text: '备忘录模式',
          exampleName: '游戏存，读档案例',
        },
        {
          linkTo: RouterEnum.observer,
          Icon: 'eye',
          text: '观察者模式',
          exampleName:
            '订阅漫画更新案例、气象站发布新消息案例、诸多公告牌进行更新案例',
        },
        {
          linkTo: RouterEnum.status,
          Icon: 'alert',
          text: '状态模式',
          exampleName: '打工人不同时段工作状态案例',
        },
        {
          linkTo: RouterEnum.strategy,
          Icon: 'ordered-list',
          text: '策略模式',
          exampleName: '购物车折扣计算案例',
        },
        {
          linkTo: RouterEnum.template,
          Icon: 'copy',
          text: '模板模式',
          exampleName: '学生抄写试题案例',
        },
        {
          linkTo: RouterEnum.interpreter,
          Icon: 'solution',
          text: '解释器模式',
          exampleName: '读懂老板的心案例、音乐解释器案例',
        },
        {
          linkTo: RouterEnum.visitor,
          Icon: 'usergroup-add',
          text: '访问者模式',
          exampleName: '读懂老板的心案例、音乐解释器案例',
        },
      ],
    },
  },
  {
    isSubMenu: false,
    linkTo: RouterEnum.unknown,
    Icon: 'question',
    text: '未知',
  },
]
